operator to get maximal element in list

oskgo opened the pull request:
There doesn't seem to be a ```max``` operator for lists yet. This implements that. I need it for defining the sup-norm.

bgregoir reviewed (commented):
I think we should not use an abstract theory to define the operator. 
It would be better, that max take the relation has arguments and that the lemmas on it 
are parametrized by the axiom. 
You can easily do that by using section instead of abstract theory.

chdoc commented:
Is there a reason why you would want to take a relation as an argument? From the title, I would have expected something along the lines of:
```
op foldl1 f dflt (s : 'a list) = 
  with s = [] => dflt
  with s = x::s => foldl f x s.

op list_max (mx : 'a -> 'a -> 'a) (dflt : 'a) = foldl1 mx dflt.  (* this is just a renaming ... *)
```

strub commented:
I would simply use an `argmax` here.

oskgo commented:
@bgregoir I'll change this. The reason I used an abstract theory is to make it easier to use if you need to use it more than once.

@chdoc I'm taking in a relation because that makes it more convenient to state, and probably prove, the axioms.

@strub Using `argmax` might shorten the proofs, but would stop the operator from being computable. Do you want me to change this?

bgregoir commented:
For me it is now ok. I propose to merge it

strub commented:
@oskgo could you squash the history with the commit message you want and we are good to go.

oskgo commented:
@strub I've squashed the history. Note that I did a mistake with git and merged the main branch into my PR at some point. I've fixed this, but it makes the final comparison look awful because the previous commit has all the changes from main in it.

bgregoir commented:
Can we merge, or people still have complaints ?
